app/process_proposal.go

```diff
@@ -66,19 +67,42 @@ func (app *App) ProcessProposal(req abci.RequestProcessProposal) (resp abci.Resp
 
 		// handle non-blob transactions first
 		if !isBlobTx {
-			_, has := hasPFB(sdkTx.GetMsgs())
+			msgs := sdkTx.GetMsgs()
+
+			_, has := hasPFB(msgs)
 			if has {
 				// A non blob tx has a PFB, which is invalid
 				logInvalidPropBlock(app.Logger(), req.Header, fmt.Sprintf("tx %d has PFB but is not a blob tx", idx))
 				return reject()
 			}
 
+			if appVersion, ok := upgrade.IsUpgradeMsg(msgs); ok {
+				if idx != 0 {
+					logInvalidPropBlock(app.Logger(), req.Header, fmt.Sprintf("upgrade message %d is not the first transaction", idx))
+					return reject()
+				}
+
+				if !IsSupported(appVersion) {
+					logInvalidPropBlock(app.Logger(), req.Header, fmt.Sprintf("block proposes an unsupported app version %d", appVersion))
+					return reject()
+				}
+
+				// app version must always increase
+				if appVersion <= app.GetBaseApp().AppVersion() {
+					logInvalidPropBlock(app.Logger(), req.Header, fmt.Sprintf("block proposes an app version %d that is not greater than the current app version %d", appVersion, app.GetBaseApp().AppVersion()))
+					return reject()
+				}
```

Comment:
[comment/discussion, no changes needed]

imo I think preserving the ability to downgrade in scenarios that don't involve could still be super useful ref https://github.com/celestiaorg/celestia-app/issues/2464

if we wanted to be super careful, we could even require that all upgrades must also be revertable including ones that require migrations. Then we could revert in the advent of an unforseen bug. This would certainly add a testing burden, but potentially worth it.

Reply:
Yeah I thought about downgrades a bit and I think there are a few things we can add to keep the software robust. My concern at the moment is that someone could just propose a downgrade and everyone would accept. 

I think we would need a second signalling mechanism which signals our willingness to accept a downgrade proposal. 

Maybe we can make an issue for this and follow up later before the PR gets way too big

Reply:
yeah I agree, downgrading (or incrementing the app version that effictively is lower verison) is far outside the scope of this PR.